Sources/CacheAdvance/FileHandleExtensions.swift

```diff
@@ -0,0 +1,160 @@
+        case .emptyRead:
+            // Seek back to the beginning of the file.
+            try seek(toOffset: 0)
+
+            if startingOffset != 0 {
+                // We hit an empty read at the end of the file.
+                // We know there's a message to read now that we're at the start of the file.
+                return try nextEncodedMessage(cacheOverwritesOldMessages: cacheOverwritesOldMessages)
+            }
+            return nil
+
+        case .invalidFormat:
+            throw CacheAdvanceReadError.fileCorrupted
+        }
+    }
+
+    /// Seeks just ahead of the newest message in the file.
+    /// - Parameter cacheOverwritesOldMessages: When `true`,  the cache encodes a pointer to the oldest message after the newest message marker.
+    func seekToEndOfNewestMessage(cacheOverwritesOldMessages: Bool) throws {
+        while try seekToNextMessage(shouldSeekToOldestMessageIfFound: false, cacheOverwritesOldMessages: cacheOverwritesOldMessages) {}
+    }
+
+    /// Seeks to the beginning of the oldest message in the file.
+    /// - Parameter cacheOverwritesOldMessages: When `true`,  the cache encodes a pointer to the oldest message after the newest message marker.
+    func seekToBeginningOfOldestMessage(cacheOverwritesOldMessages: Bool) throws {
+        if cacheOverwritesOldMessages {
+            while try seekToNextMessage(shouldSeekToOldestMessageIfFound: true, cacheOverwritesOldMessages: true) {}
+        } else {
+            // The oldest message is always at the beginning of the cache.
+            try seek(toOffset: 0)
+        }
+    }
+
+    /// Seeks to the next message. Returns `true` when the span skipped represented a message.
+    /// When `false` is returned, it signifies that the last message marker was passed.
+    ///
+    /// - Parameters:
+    ///   - shouldSeekToOldestMessageIfFound: When `true`, the file handle will seek to the oldest message if the last message marker is pased.
+    ///   - cacheOverwritesOldMessages: When `true`,  the cache encodes a pointer to the oldest message after the newest message marker.
+    @discardableResult
+    func seekToNextMessage(shouldSeekToOldestMessageIfFound: Bool, cacheOverwritesOldMessages: Bool) throws -> Bool {
+        let startingOffset = offsetInFile
+        switch try nextEncodedMessageSpan(cacheOverwritesOldMessages: cacheOverwritesOldMessages) {
+        case let .endOfNewestMessageMarker(offsetOfNextMessage):
+            // We found the last message!
+            if shouldSeekToOldestMessageIfFound {
+                // Seek to it.
+                try seek(toOffset: offsetOfNextMessage)
+            } else {
+                // Seek back to where we started.
+                try seek(toOffset: startingOffset)
+            }
+            return false
+
+        case .emptyRead:
+            // The remaining file is empty.
+            if shouldSeekToOldestMessageIfFound {
+                // The oldest message is at the beginning of the file.
```

Comment:
Do we only get back an `.emptyRead` if we have a cache that will never overwrite old messages?
If we can get back `.emptyRead` for a cache that _does_ overwrite old message, it seems like we can't assume the oldest message is at the start of the file.

Reply:
If this behavior is expected, maybe there is a way that we can clarify.

Reply:
The only way we can get an `emptyRead` is if:
```
// We haven't written anything to this file yet, or we've reached the end of the file.
```

I think given that comment I can simplify this code.

Reply:
Looking at this again, I think you're right. In a cache that overrides old messages, we'll hit the oldest message offset marker before we have an empty read – getting an `emptyRead` in that case would likely mean there was some kind of programmer error (we'd have to seek to the next message span after reading the oldest message offset marker). I'll clarify with a comment.